### DISCUSSION TITLE
Connection issue

### DISCUSSION THREAD

**Discussion by elim17** [USC-ACTLab/crazyswarm#332] [Q&A]
Based on what was said on this page [https://github.com/USC-ACTLab/crazyswarm/issues/32](url) I should not have to run a vrpn client node to connect to motive Optitrack. When I do not, however, nothing shows up in Rviz. We are running Optitrack on its own workstation that has windows. On a separate workstation that runs linux, we are running the crazy swarm software. Any ideas on what is causing the disconnect when not using vrpn? Thank you!

**Comment by whoenig** [comment]
For Optitrack related issues, please verify that NatNet works for you, using the following repository: https://github.com/whoenig/NatNetSDKCrossplatform. The Crazyswarm internally simply uses that SDK for communication with OptiTrack.

**Reply by elim17** [reply]
Thank you. I ran the natnettest and mcr and both seem to have run property. I am still unable to see the crazyflie in Rviz without running the vrpn client. Is there anything else that could be causing this problem. Do I need to manually open NatNet before running hover.launch?

**Reply by whoenig** [reply]
No need to run anything else. The next step is to run `mocap_helper.launch`. In all cases, make sure that the launch file is configured correctly with the motion capture type and hostname/IP of the machine that runs Motive.

**Reply by elim17** [reply]
Ok. When I run mocap_helper.launch it seems to run alright. Does this output look alright to you? 
When I run the hover_swarm.launch file I still do not see anything in Rviz. 
```
15:41:28|~/crazyswarm$ source ros_ws/devel/setup.bash
removed '/tmp/setup.sh.OyDqWL8Qsa'
15:42:42|~/crazyswarm$ roslaunch crazyswarm mocap_helper.launch
... logging to /home/hcrobotics/.ros/log/c624d9bc-77ba-11eb-9523-40b0769f73bd/roslaunch-hcrobotics-6718.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://192.168.1.100:44147/

SUMMARY
========

PARAMETERS
 * /mocap_helper/motion_capture_type: optitrack
 * /mocap_helper/object_tracking_type: motionCapture
 * /mocap_helper/optitrack_local_ip: 192.168.1.100
 * /mocap_helper/optitrack_server_ip: 192.168.1.69
 * /rosdistro: melodic
 * /rosversion: 1.14.10

NODES
  /
    mocap_helper (crazyswarm/mocap_helper)

auto-starting new master
process[master]: started with pid [6728]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to c624d9bc-77ba-11eb-9523-40b0769f73bd
process[rosout-1]: started with pid [6739]
started core service [/rosout]
process[mocap_helper-2]: started with pid [6746]
```
This is my mocap_helper.launch file if that helps. For Optitrack do we need the hostname and IP's or just the IP's? 

```
<?xml version="1.0"?>
<launch>

  <node pkg="crazyswarm" type="mocap_helper" name="mocap_helper" output="screen" >
    <rosparam>
      # tracking
      motion_capture_type: "optitrack" # one of vicon,optitrack,qualisys,vrpn
      object_tracking_type: "motionCapture" # one of motionCapture,libobjecttracker
      #vicon_host_name: "vicon"
      #optitrack_host_name: "optitrack"
      # qualisys_host_name: "10.0.5.219"
      # qualisys_base_port: 22222
      #vrpn_host_name: "192.168.1.69"
      optitrack_local_ip: "192.168.1.100"
      optitrack_server_ip: "192.168.1.69"
    </rosparam>
  </node>
</launch>
```

**Reply by whoenig** [reply]
Just the optitrack_host_name is used in the current code. There should be a print out of markers and/or rigid bodies on the terminal if it is working correctly, similar to `mcr`.

**Reply by elim17** [reply]
@whoenig I changed the launch files to just use optitrack_host_name set to our Optitrack server IP address, but I still do not have any print out like that of `mcr` when I run mocap_helper.launch. Is there a flag I need to use to get the print out of markers?

**Comment by lkyfly0765** [comment]
Please make sure that when you run the ./mcr, the output contains the position information of your rigid body, and double check whether they are correct or not.

It seems weird that your ./mcr runs properly but have problems in Rviz. Please try to run the ./mcr program in another terminal window and then `roslaunch` your launch file, this should work to solve your Rviz issue.

Your mobcap_helper.launch is properly configured according to your information.

**Reply by elim17** [reply]
Thank you. The ./mcr runs correctly and gives the position information fo the rigid body. They do update when the rigid body moves. I ran the ./mcr program in another terminal window then tried running both mocap_helper.launch and hover_swarm.launch.  The rigid body did not show up in Rviz still even though the ./mcr was tracking just fine. I can get the output of hover_swarm.launch later today, but below is the output of a packet from ./mcr and the output of running mocap_helper.launch. Note: mobcap_helper.launch still does not print out data like ./mcr. Both are running at the same time in two different terminals. 
![Screenshot from 2021-03-02 14-16-52](https://user-images.githubusercontent.com/42954721/109835727-b8f76f80-7c00-11eb-8823-c7c98d3e1e76.png)

![Screenshot from 2021-03-02 14-08-18](https://user-images.githubusercontent.com/42954721/109836376-52268600-7c01-11eb-9fef-620b873ba032.png)

**Reply by whoenig** [reply]
There is no need to configure anything. Make sure that your libmotioncapture is on the latest version (did you run "git submodule update"?). You can also add some manual print/cout statements in https://github.com/USC-ACTLab/libmotioncapture/blob/master/src/optitrack.cpp, to see where it hangs.

**Reply by elim17** [reply]
It would seem that the changes I made in my launch files were not represented in the test that I did yesterday.The change was from using Optitrack_local_ip and Optitrack_server_ip to just using Optitrack_host_name. I ran it again today and received promising results as well as an error that puts me at a loss. 

I ran ./mcr <ip> in its own terminal(output here looks good. Can see the drone move in the cordinates). When I ran mocap_helper.launch consecutively this is the output:

![Screenshot from 2021-03-03 14-32-04](https://user-images.githubusercontent.com/42954721/109879278-9d0cc180-7c32-11eb-8b24-884521962ea0.png)

The promising part is that you can see that it started to run something similar to ./mcr where it prints `frame` 0. It wasn't doing that before. If you could give any suggestions on how to handle this error that would be great. Thank you.

**Reply by whoenig** [reply]
OK - so it looks like it receives data, but it's not parsing it correctly. Which Motive version are you using? We only tested with 2.X. Also, some users reported that you have to only enable what you need in the streaming pane in motive. For example, if you only want to get the raw point cloud, only enable "unlabeled markers" in Motive to be streamed out.

**Reply by whoenig** [reply]
The "proper" fix will be to go over the parsing code in https://github.com/USC-ACTLab/libmotioncapture/blob/master/src/optitrack.cpp#L333-L630, and make sure it semantically matches the code in mcr. Unfortunately, all of this code is super ugly (not our fault, that's just how the OptiTrack SDK is...)

**Reply by elim17** [reply]
We are using Motive: Tracker 2.2.0. Could you explain more about raw point cloud vs rigid body? Our first goal is to be able to run hello_world.py and get one drone to hover.

**Reply by whoenig** [reply]
There is some documentation here: https://crazyswarm.readthedocs.io/en/latest/configuration.html#adjust-configuration-files. You essentially have to decide if you want Motive to track your object or libobjecttracker. I can only recall testing the latter with Optitrack myself. In either case, make sure that in the Motive Streaming pane you disabled the part of the stream that you do not want, because there seems to be some decoding bug in libmotioncapture that causes the crash.

**Reply by elim17** [reply]
Our lab decided to go with Motive to track our rigid bodies. Your point about parsing was correct. We were receiving too much data from Motive via NatNet. By turning off the point cloud data mocap_helper.launch was able to run correctly! NOTE: we did not need t have ./mcr running at the same time for it to work. Thank you for the help!
